File: dotnet/Platforms/GameX.Platform.OpenGL/OpenGL_Render.cs
using GameX.Gamebryo.Formats;
using OpenStack;
using OpenStack.Gfx;
using OpenStack.Gfx.OpenGL;

namespace GameX.Platforms.OpenGL;

public static class OpenGLRenderer {
    static OpenGLRenderer() {
        OpenGLPlatform.BuildersByType[typeof(Binary_Nif)] = OpenGLNifObjectBuilder.BuildObject;
    }
    public static Renderer CreateRenderer(object parent, IOpenGfx[] gfx, ISource source, object obj, string type) {
        if (obj is IHaveSource z) source = z.Source;
        return type switch {
            "TestTri" => new TestTriRenderer(gfx, source, obj),
            "Texture" or "VideoTexture" => new TextureRenderer(gfx, source, obj, 0.., false),
            "Object" => new ObjectRenderer(gfx, source, obj),
            "Material" => new MaterialRenderer(gfx, source, obj),
            "Particle" => new ParticleRenderer(gfx, source, obj),
            //"World" => new OpenGLWorldRenderer(gfx, source,obj),
            "Engine" => new EngineRenderer(gfx, source, obj),
            _ => default,
        };
    }
}

File: python/gamex/platforms/opengl/opengl_render.py
from __future__ import annotations
import os
from openstk.core import ISource, IHaveSource
from openstk.platforms.opengl.gfx import TestTriRenderer, TextureRenderer, ObjectRenderer, EngineRenderer

# typedefs
class IOpenGfx: pass
class Renderer: pass

# OpenGLRenderer
class OpenGLRenderer:
    @staticmethod
    def createRenderer(parent: object, gfx: list[IOpenGfx], source: ISource, obj: object, type: str) -> Renderer:
        if isinstance(obj, IHaveSource): source = obj.source
        match type:
            case 'TestTri': return TestTriRenderer(gfx, source, obj)
            case 'Texture' | 'VideoTexture': return TextureRenderer(gfx, source, obj, None, False)
            case 'Object': return ObjectRenderer(gfx, source, obj)
            # case 'Material': return MaterialRenderer(gfx, source, obj)
            # case 'Particle': return ParticleRenderer(gfx, source, obj)
            case 'Engine': return EngineRenderer(gfx, source, obj)
            # case 'World': return WorldRenderer(gfx, source, obj)
            case _: return None
